lib/booking_manager.rb

```diff
@@ -0,0 +1,184 @@
+require_relative 'room'
+require_relative 'reservation'
+# require 'pry'
+
+module Hotel
+  class BookingManager
+    attr_accessor :rooms, :reservations, :room_calendar
+
+    def initialize(number_rooms)
+      @rooms = populate_room_list(number_rooms) #(20)
+      @reservations = make_reservation_list
+      @room_calendar = make_room_calendar(number_rooms)
+    end
+
+    # Create list of rooms as list of room Instances
+    def populate_room_list(number_rooms)
+      rooms = []
+      num = 1
+
+      number_rooms.times do |room|
+        room = Room.new(num)
+        #
+        rooms << room
+        num += 1
+      end
+
+      return rooms
+    end
+
+
+    # Create array to store all of reservations
+    def make_reservation_list
+      reservations = []
+      return reservations
+    end
+
+
+    # Create list of rooms with reserved dates.
+    def make_room_calendar(number)
+      @room_calendar = {}
+
+      @rooms.each do |room|
+        dates_reserved = {}
+        @room_calendar[room] = dates_reserved
+      end
+
+      return @room_calendar
+    end
+
+
+    # Method to add a reservation to list of reservations
+    def add_reservation(reservation)
+      @reservations << reservation
+    end
+
+
+    # Add reservation date range to hash of room reserved dates
+    def add_reservation_to_calendar(reservation)
+      date = reservation.start_date
+
+      reservation.number_nights.times do
+        @room_calendar[reservation.room][date] = reservation # dependency
+        date += 1
+      end
+# binding.pry
+      return @room_calendar
+    end
+
+
+    # Check if date range given is valid - start must be before end
+    def check_dates(start_date, end_date)
+      if start_date > end_date
+        raise ArgumentError.new "Invalid date range. Start date must be before end date, both in format of 'Month dd, yyyy'. "
+      end
+    end
+
+
+    # Create array of all dates from start date to end date
+    def determine_date_range(start_date, end_date)
+      check_dates(start_date, end_date)
+
+      res_start_date = Date.parse(start_date)
+      res_end_date = Date.parse(end_date)
+
+      date_range = []
+      search_date = res_start_date
+
+      until search_date > res_end_date
+        date_range << search_date
+        search_date += 1
+      end
+      return date_range
+    end
+
+    # Method to list all reservation instances
+    def list_reservations
+      return @reservations
+    end
+
+
+    # Method to list all rooms in hotel
+    def list_rooms
+      return @rooms
+    end
+
+
+    # Method to get total cost of reservation
+    def get_reservation_cost(nights, cost_per_night)
+      total_cost = nights * cost_per_night
+      return total_cost
```

Comment:
The `Reservation` should have all the info about the number of nights and the cost per night, which indicates to me that this should be a method on the `Reservation`, not on the `BookingManager`. This is an example of not quite giving `Reservation` a whole responsibility (which in turn means `BookingManager` has more than one responsibility).